crates/rattler_repodata_gateway/src/gateway/query.rs

```diff
@@ -92,9 +95,30 @@ impl GatewayQuery {
             .cartesian_product(self.platforms.into_iter())
             .collect_vec();
 
-        // Create barrier cells for each subdirectory. This can be used to wait until the subdir
-        // becomes available.
-        let mut subdirs = Vec::with_capacity(channels_and_platforms.len());
+        // Package names that we have or will issue requests for.
+        let mut seen = HashSet::new();
+        let mut pending_package_specs = HashMap::new();
+        let mut direct_url_specs = vec![];
+        for spec in self.specs {
+            if spec.url.is_some() {
+                let name = spec
+                    .name
+                    .clone()
+                    .ok_or(GatewayError::MatchSpecNoName(spec.clone()))?;
+                seen.insert(name.clone());
+                direct_url_specs.push(spec.clone());
```

Comment:
I dont think this clone is required right?

```suggestion
                direct_url_specs.push(spec);
```